packages/core/src/indexing/service.ts

```diff
@@ -527,264 +712,277 @@ export class IndexingService extends Emittery<IndexingEvents> {
-
-                this.common.metrics.ponder_indexing_has_error.set(1);
-                this.emit("error", { error });
-              } else {
-                this.common.logger.warn({
-                  service: "indexing",
-                  msg: `Indexing function failed, retrying... (event=${fullEventName}, block=${Number(
-                    event.block.number,
-                  )}, error=${`${error.name}: ${error.message}`})`,
-                });
-                await this.indexingStore.revert({
-                  checkpoint: this.currentIndexingCheckpoint,
-                });
-              }
-            }
-          }
-
+          await this.executeLogEventTask(task);
           break;
         }
       }
     };
 
-    const queue = createQueue({
+    this.queue = createQueue({
       worker: indexingFunctionWorker,
       options: {
-        concurrency: 1,
+        concurrency: 10,
         autoStart: false,
       },
     });
+  };
+
+  /**
+   * Load a batch of indexing function tasks from the sync store into memory.
+   */
+  loadIndexingFunctionTasks = async (key: string) => {
+    const keyHandler = this.indexingFunctionMap![key];
+    const tasks = keyHandler.indexingFunctionTasks;
+
+    if (
+      tasks.length > 0 ||
+      isCheckpointEqual(
+        keyHandler.maxTaskCheckpoint,
+        this.syncGatewayService.checkpoint,
+      )
+    )
+      return;
+
+    const { events, metadata } = await this.syncGatewayService.getEvents({
+      fromCheckpoint: keyHandler.maxTaskCheckpoint,
+      toCheckpoint: this.syncGatewayService.checkpoint,
+      limit: TASK_BATCH_SIZE,
+      logFilters: keyHandler.sources
+        .filter(sourceIsLogFilter)
+        .map((logFilter) => ({
+          id: logFilter.id,
+          chainId: logFilter.chainId,
+          criteria: logFilter.criteria,
+          fromBlock: logFilter.startBlock,
+          toBlock: logFilter.endBlock,
+          includeEventSelectors: [keyHandler.eventSelector],
+        })),
+      factories: keyHandler.sources.filter(sourceIsFactory).map((factory) => ({
+        id: factory.id,
+        chainId: factory.chainId,
+        criteria: factory.criteria,
+        fromBlock: factory.startBlock,
+        toBlock: factory.endBlock,
+        includeEventSelectors: [keyHandler.eventSelector],
+      })),
+    });
+
+    if (events.length < TASK_BATCH_SIZE) {
+      keyHandler.maxTaskCheckpoint = this.syncGatewayService.checkpoint;
+    } else {
+      keyHandler.maxTaskCheckpoint = metadata.endCheckpoint;
+    }
+
+    // Update handled and matched events, unawaited because not in critical path.
+    this.setEventMetrics();
+
+    const abi = [keyHandler.abiEvent];
+    const contractName = keyHandler.contractName;
+    const eventName = keyHandler.eventName;
+
+    for (const event of events) {
+      try {
+        const decodedLog = decodeEventLog({
+          abi,
+          data: event.log.data,
+          topics: event.log.topics,
+        });
```

Comment:
Random thought - I wonder if there's any perf gain to be had by decoding all `events` first, then pushing them to the queue. Decoding could also happen in a Worker 😈